In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime

In [ ]:
source_data = pd.read_csv('/content/cars_rawdata.csv')

In [ ]:
source_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

Manage missing values and eliminate units from the columns (a&b).

In [ ]:
# Check for lost values in each column
lost_values = source_data.isnull().sum()
print("Lost Values")
print(lost_values)

Lost Values
Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


#### **By above observation, Mileage, Engine, Power, Seats & New_Price has null or missing values.**

### **1. Handle Mileage column**


The error arises from the inability to convert '19.67 kmpl' to numeric format and the solution is to remove the units.

In [ ]:
 #Check for lost values in the 'Mileage' column before handling
print("Before Handling, Number of Lost Values in Mileage Column:", source_data['Mileage'].isnull().sum())

# Regular Expression is used in str.extract() function to extract substrings
# \d+: Matches one or more digits.
# \.: Matches a literal dot (period).
# \d+: Matches one or more digits after the dot.
# astype(float) function is then used to convert the extracted substrings (str) to float.
source_data['Mileage'] = source_data['Mileage'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Since Mileage is reported in kmpl,
# it's reasonable to assume that missing values are likely to be missing at random.
# In such cases, filling with the median is a common strategy.
source_data['Mileage'].fillna(source_data['Mileage'].median(), inplace=True)

# Check for Lost values in the 'Mileage' column after handling
print("After Handling, Number of Lost Values in Mileage Column:", source_data['Mileage'].isnull().sum())


Before Handling, Number of Lost Values in Mileage Column: 2
After Handling, Number of Lost Values in Mileage Column: 0


### **2. Handling the Engine column**


The error indicates that '300 CC' cannot be converted to numeric format, and the solution is to remove the units.

In [ ]:
# Check for Lost values in the 'Engine' column before handling
print(" Number of Lost Values in Engine Column before handling is:", source_data['Engine'].isnull().sum())

# Remove units

source_data['Engine'] = source_data['Engine'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Lost values for engine size are often handled by filling with the mean value
source_data['Engine'].fillna(source_data['Engine'].mean(), inplace=True)

# Check for Lost values in the 'Engine' column after handling
print(" Number of Lost Values in Engine Column after handling is:", source_data['Engine'].isnull().sum())

 Number of Lost Values in Engine Column before handling is: 36
 Number of Lost Values in Engine Column after handling is: 0


### **3. Handling the Power Column**

The error represents that sample'39.09 bhp' cannot convert to numeric and to solve this we need to Remove the units.

In [ ]:
# Check for lost values in the 'Power' column before handling
print("Number of Missing Values in Power Column before handling is:", source_data['Power'].isnull().sum())

# Removing of units
source_data['Power'] = source_data['Power'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Lost values for Power are often handled by filling with the mean value
source_data['Power'].fillna(source_data['Power'].mean(), inplace=True)

# Check for Lost values in the 'Power' column after handling
print("Number of Missing Values in Power Column after handling is:", source_data['Power'].isnull().sum())

Number of Missing Values in Power Column before handling is: 36
Number of Missing Values in Power Column after handling is: 0


### **4. Handling the Seats Column**

In [ ]:
# Check for Lost values in the 'Seats' column before handling
print("Number of Lost Values in Seats Column before handling", source_data['Seats'].isnull().sum())

# lost values for seats of the car are often handled by filling with the mean value
source_data['Seats'].fillna(source_data['Seats'].mean(), inplace=True)

# Check for lost values in the 'Seats' column after handling
print("Number of Lost Values in Seats Column after handling", source_data['Seats'].isnull().sum())


Number of Lost Values in Seats Column before handling 38
Number of Lost Values in Seats Column after handling 0


### **5. Handling the New_Price Column**

In [ ]:
# Drop the New_Price column since it has a high number of missing values
source_data.drop('New_Price', axis=1, inplace=True)

**View the no null values handled DataFrame.**

In [ ]:
missing_values = raw_data.isnull().sum()
missing_values

Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 0
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64

# `(c)  Change the categorical variables (“Fuel_Type” and “Transmission”) into numerical one hot encoded value`

In [ ]:
# Encoding Categorical Variables by using get_dummies function from pandas
source_data = pd.get_dummies(source_data, columns=['Fuel_Type', 'Transmission'])

# Showing the updated dataset
print(source_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

# `(d)  Create one more feature and add this column to the dataset. `

### **Creating age_of_car feature which define the age of the car.**

In [ ]:
# Calculating the current year
current_year = datetime.now().year

# Minus the "Year" column from the current year to find out the age of the car
source_data['Age_of_car'] = current_year - source_data['Year']

# Display the new data set
print(source_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

#### **Creating Mileage_Category feature to define mileage value by category.**

In [ ]:
# Define the mileage categories
m_b = [0, 10, 20, 30, float('inf')]
m_l = ['Very Low', 'Low', 'Medium', 'High']

# Create a new feature 'Mileage_Category' based on the mileage groups
source_data['Mileage_range'] = pd.cut(source_data['Mileage'], bins=m_b, labels=m_l, right=False)

# Display the modified dataset with the new feature
print(source_data.head())


   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

# `(e) Perform select, filter, rename, mutate, arrange and summarize with group by operations (or their equivalent operations in python) on this dataset.`

In [ ]:
# Selecting required columns
selected_data = source_data[['Name', 'Location', 'Year', 'Mileage', 'Price','Age_of_car']]
# Displaying the results
print("Selected Data:")
print(selected_data.head())

Selected Data:
                               Name    Location  Year  Mileage  Price  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015    19.67  12.50   
1                      Honda Jazz V     Chennai  2011    13.00   4.50   
2                 Maruti Ertiga VDI     Chennai  2012    20.77   6.00   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013    15.20  17.74   
4            Nissan Micra Diesel XV      Jaipur  2013    23.08   3.50   

   Age_of_car  
0           9  
1          13  
2          12  
3          11  
4          11  


In [ ]:
# Filtering the data based on year
filtered_data = source_data[source_data['Year'] >= 2017]
print("\nFiltered Data (Year >= 2019):")
print(filtered_data.head())


Filtered Data (Year >= 2019):
    Unnamed: 0                                     Name Location  Year  \
8           10                         Maruti Ciaz Zeta    Kochi  2018   
15          17                    Maruti Swift DDiS VDI   Jaipur  2017   
26          28                      Honda WRV i-VTEC VX    Kochi  2018   
34          36            Maruti Vitara Brezza ZDi Plus    Kochi  2018   
35          37  Mahindra KUV 100 mFALCON G80 K6 5str AW    Delhi  2017   

    Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
8               25692      First    21.56  1462.0  103.25    5.0   9.95   
15              25000      First    28.40  1248.0   74.00    5.0   5.99   
26              37430      First    17.50  1199.0   88.70    5.0   9.90   
34              50075      First    24.30  1248.0   88.50    5.0   8.63   
35              52000      First    18.15  1198.0   82.00    5.0   4.85   

    Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
8          

In [ ]:
# Renaming columns
renamed_data = source_data.rename(columns={'Engine': 'Engine_Measurment', 'Power': 'strength_of_car_engine'})

print("\nRenamed Columns:")
print(renamed_data.head())


Renamed Columns:
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine_Measurment  \
0              41000      First    19.67             1582.0   
1              46000      First    13.00             1199.0   
2              87000      First    20.77             1248.0   
3              40670     Second    15.20             1968.0   
4              86999      First    23.08             1461.0   

   strength_of_car_engine  Seats  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
0                  126.20    5.0  12.50                 1                   0   
1                   88.70 

In [ ]:
# Mutating data
source_data['Price_Category'] = pd.cut(source_data['Price'], bins=[0, 5, 10, 15, 20, 25, 30], labels=['<5', '5-10', '10-15', '15-20', '20-25', '25-30'])

print("\nMutated Data (Added Price_Category column):")
print(source_data.head())


Mutated Data (Added Price_Category column):
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1               

In [ ]:
#Arranging data based on the year in descending order:
arranged_data = source_data.sort_values(by='Year', ascending=False)

print("\nArranged Data (Sorted by Year in descending order):")
print(arranged_data.head())


Arranged Data (Sorted by Year in descending order):
      Unnamed: 0                             Name    Location  Year  \
5405        5560        Renault KWID RXT Optional       Kochi  2019   
942          975  Ford Endeavour 2.2 Trend AT 4X2       Kochi  2019   
5533        5690        Maruti Omni 5 Seater BSIV  Coimbatore  2019   
770          796            Mahindra XUV500 W9 AT  Coimbatore  2019   
4267        4399       Maruti Swift Dzire AMT ZDI     Chennai  2019   

      Kilometers_Driven Owner_Type  Mileage  Engine  Power  Seats  Price  \
5405               6568      First    25.17   799.0   53.3    5.0   5.09   
942               11209      First    12.62  2198.0  158.0    7.0  31.15   
5533               4721      First    14.00   796.0   35.0    5.0   4.11   
770               19654      First    14.00  2179.0  155.0    7.0  17.63   
4267              65000      First    26.59  1248.0   74.0    5.0   6.75   

      Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \

In [ ]:
# Summarizing data with group by operations
summary_data = source_data.groupby('Location')['Price'].mean()

print("\n(Average Price by Location):")
print(summary_data)


(Average Price by Location):
Location
Ahmedabad      8.567248
Bangalore     13.482670
Chennai        7.958340
Coimbatore    15.160206
Delhi          9.881944
Hyderabad      9.997423
Jaipur         5.916725
Kochi         11.309109
Kolkata        5.733924
Mumbai         9.592546
Pune           6.951000
Name: Price, dtype: float64


# `Save source_data(DataFrame) in New CSV`

In [ ]:
source_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 0                   0                 1   
2                 1                   0                 0   
3                 1                   0                 0   
4                 1                   0                 0   

   Transmission_Automatic  Transmission_Manual  Age_of_car Mileage_range  \
0                       0                    1           9           Low   
1                       0                    1          13           Low   
2                       0                    1          12        Medium   
3                       1                    0          11           Low   
4                       0                    1          11        Medium   

  Price_Category  
0          10-15  
1             <5  
2           5-10  
3          15-20  
4             <5

In [ ]:
#Save the data into clean data csv
source_data.to_csv("/content/clean_data.csv")

In [ ]:
from google.colab import drive
drive.mount('/content/drive')